README

(This file was last updated November 18, 2013 by Pete Siemsen)

This file describes a set of Perl scripts called "SwitchMap" that
generate handy Web pages describing Cisco switches.  The comments at
the top of the SwitchMap.pl file contain more information.


INSTALLING SwitchMap on Microsoft Windows:

See Windows_Installation.pdf for Windows-specific instructions.  Even
if you read that file, you'll probably want to read this file anyway
to learn how SwitchMap works.


INSTALLING SwitchMap on Unix/Linux:

Below, I've tried to guess what will be needed to get the script
working at your site.  It's assumed that you have Perl 5.0.  If there
are problems, please send email to the address given at the bottom of
this file.


INSTALLING
1. Get the Perl modules needed by switchmap.  If you are running
   Debian Linux, this should do it:

   (as root)
   apt-get install libnet-snmp-perl
   apt-get install liblog-log4perl-perl
   apt-get install liblog-dispatch-perl
   apt-get install libsnmp-perl

   If you have a non-Debian system, with Perl installed, try the
   following:

   (as root)
   perl -MCPAN -e shell
   install Log::Log4perl
   install Module::Build
   install Net::SNMP
   install Log::Dispatch::Screen
   h
   q


2. Unpack the distribution

   The switchmap Perl scripts create web pages, so you'll want to
   install the scripts on a machine that has a web server.  Put the
   distribution file (named something like switchmap-12.5.tar.gz) in a
   directory somewhere on the machine.  If you have permisson, put it
   where your operating system likes to store third-party software.
   On Unix systems, this is /usr/local.  Gunzip/untar the .gz file.
   You should end up with a directory named something like
   "switchmap-12.5".  It should contain the following files:

    CHANGELOG                      Shows history of changes
    CISCO-ENTITY-VENDORTYPE-OID-MIB.my  From Cisco, provides SNMP OIDs
    CISCO-PRODUCTS-MIB.my          From Cisco, provides SNMP OIDs
    CISCO-STACK-MIB.my             From Cisco, provides module data
    CiscoMibConstants.pm           functions to read Cisco MIB constants
    Constants.pm                   Constant definitions
    EtherChannel.pm                EtherChannel class
    FindOffice.pl                  CGI script, implements "search"
    GetArp.pl                      Program to get ARP caches from routers
    GetCiscoPortTrunkStatuses.pm   Gets some port data from switches
    GetMacsFromCiscoMibs.pm        Gets MAC addresses from Cisco CAM tables
    GetMacsFromQBridgeMib.pm       Gets MAC addresses from the Q-Bridge MIB
    GetVlansOnPorts.pm             Gets VLANs on trunk ports
    hpMibConstants.pm              functions to read HP MIB constants
    MacIpTables.pm                 Gets MAC addresses, hostnames, switchnames
    MibConstants.pm                functions to read MIB constants
    ModuleList.pm                  ModuleList class
    OuiCodes.pm                    OuiCodes hash
    OuiCodes.txt                   Organizationally Unique Identifiers table
    PetesUtils.pm                  General utility functions
    PopulateEtherChannels.pm       Contains the PopulateEtherChannels function
    PopulatePorts.pm               Contains the PopulatePort function
    Port.pm                        Port class
    Portically.pm                  contains the "portically" sorting function
    README                         this file
    ScanSwitch.pl                  Perl script to generate "idlesince" files
    SearchPortlists.html           HTML file that invokes FindOffice.pl
    SnmpCommunities.pm             SNMP communities and code to read them
    Stats.pm                       Code to produce the statistics page
    Switch.pm                      Switch class
    SwitchMap.css                  CSS file used by the generated web pages
    switchmap                      Optional Unix startup file for /etc/init.d
    switchmap-daemon               Optional script used by "switchmap"
    SwitchMap.pl                   Main program
    SwitchUtils.pm                 Include file containing shared functions
    ThisSite.pm                    Site-specific constant definitions
    UpdateOuiCodes.pl              Program to generate OuiCodes.txt
    VerifyPortLabels               Check port labels (used only at NCAR)
    Vlan.pm                        Vlan class
    Windows_Installation.pdf       How to install SwitchMap on Windows systems
    WriteCsvDirectory.pm           Writes the csv directory
    WriteGigePerVlansDirectory.pm  Writes the gigeportspervlan directory
    WriteModulesFile.pm            Writes the modulesbyswitch.html file
    WriteNcarFiles.pm              Writes files used only at NCAR
    WritePLookupDirectory.pm       Writes files used by the plookup program
    WritePortsDirectory.pm         Writes the ports directory
    WritePoePortsFile.pm           Writes the Power-over-Ethenet ports file
    WriteSuspiciousPortsFile.pm    Writes the Power-over-Ethenet ports file
    WriteSwitchesDirectory.pm      Writes the switches directory
    WriteUnusedDirectory.pm        Writes the unused directory
    WriteVlansDirectory.pm         Writes the vlans directory
    error-log.shtml                beta code to view nidex.php's error log
    index.php                      beta code to refresh SwitchMap on demand

   These files are described in more detail in the comments at the top
   of the SwitchMap.pl file.


3. Set up SNMP in your switches and routers

   Make sure your routers and Catalyst switches are configured to
   speak SNMP.  This command enables SNMP on a Cisco Catalyst switch:

     snmp community read-only public

   You should choose another SNMP community string than "public".  See
   the ThisSite.pm file for instructions for how to configure
   community strings into the switchmap programs.


4. Define site-specific variables

   Change the definitions in ThisSite.pm to fit your site.  Comments
   in the file explain this in detail.  This is where you define the
   names of the routers and switches in your network, where the output
   files should get written, etc.


5. Run GetArp.pl.

   Switchmap needs to map MAC addresses to IP addresses.  If you don't
   use HP OpenView, SwitchMap reads a mapping table from a file named
   MacList.  The MacList file is created and updated by the GetArp.pl
   script.  GetArp.pl reads ARP caches from your routers and updates
   the MacList file.  To tell GetArp.pl the names of your routers, set
   the @routers array in the ThisSite.pm file.  Router ARP cache
   entries time out, so to make MacList as complete as possible, you
   should run GetArp.pl often (see below).  If GetArp doesn't work,
   use the "-d" option to turn on debugging messages.

   As of version 12.0, SwitchMap reads ARP caches from switches as
   well as routers, so that it can produce information about routing
   ports as well as switching ports.  You still need to run GetArp.pl
   to maintain the MacList file.


6. Run ScanSwitch.pl.

   This should create a directory named "idlesince" containing a file
   for each of your Cisco switches.  If it doesn't work, use the "-d"
   option to turn on debugging messages.


7. Run SwitchMap.pl.

   This should produce HTML files in subdirectories of the
   $DestinationDirectory that you configured in your ThisSite.pm file.
   The subdirectories are named "ports", "vlans" and "switches".  If
   SwitchMap doesn't work, try using the "-d" option to turn on
   debugging messages.


8. Look at the resulting files

    Point your Web browser at the $DestinationdDirectory.  There
    should be an index.html file.  Hope it looks ok!


9. Once the above is working, you'll want to set up some mechanism to
   run the programs on a regular schedule to keep the web pages
   up-to-date.  Mechanisms to do this depend on your operating system.
   Some approaches for Unix-based systems are described here.  Windows
   users should look in the Windows_Installation.pdf file in the
   SwitchMap distribution.

    A. cron jobs

       Option 1:

    Set up one cron job to run runSwitchMap.sh, which will run all
    three SwitchMap programs.  This option is simpler than Option 2,
    but offers less flexibility - you'll tend to run SwitchMap.pl a
    little more often than you probably need, and you may run
    ScanSwitch.pl less often.  If you use this option, you'll have to
    edit runSwitchMap.sh to install some paths and such.

       Option 2:

    Set up 3 separate cron jobs to run the 3 programs regularly.  This
    is what I do at my site.  Set up one job to run ScanSwitch.pl
    every hour and one to run SwitchMap.pl every day.  If you don't
    have HP OpenView, set up a third cron job to run GetArp.pl every 3
    hours.  Of course, you may want to change these frequencies - some
    sites run SwitchMap every hour.  My crontab entries look like
    this:

#
# These entries run Pete Siemsen's switchmap scripts to gather
# information about Ethernet switch ports.  The result is web pages
# created every day that show what machines are connected to the
# switch ports.
#
# At 44 minutes after every hour, scan the switches for port
# statuses and update the .idlesince files.  At 49 minutes after
# every hour, get ARP information from the routers and update
# the MacList file.
#
44 * * * * /usr/web/nets/internal/portlists/ScanSwitch.pl
49 * * * * /usr/web/nets/internal/portlists/GetArp.pl
#
# At 2:05pm every day, during the middle of the day when the
# most ports are likely to be active, read the .idlesince files
# and create the web pages.  It's best to run this when there's

CHANGELOG
# Change log for the SwitchMap program, which is available at
# http://sourceforge.net/projects/switchmap/
#

# Version 13.1, February 5, 2014
Added the @LocalSwitchTrunkPorts array to ThisSite.pm.  This provides
  a somewhat klunky work-around for people that have certain switches
  (i.e. Junipers and Cisco 3845s).  These switches don't provide a
  way for SwitchMap to identify trunk (uplink) ports via SNMP.  For
  these switches, SwitchMap produces web pages with long lists of MAC
  addresses on those ports, instead of "trunk port".  If you have
  this problem, read the comment that accompanies
  @LocalSwitchTrunkPorts in ThisSite.default.pm.
Removed the "search networks" button, which was only meaningful
  at my site.
 Uncommented and corrected some old code to support Foundry switches.
  Still needs work.  Thanks to Andrew Struiksma for providing access
  to Foundry devices for testing.
In GetVlansOnPorts, added spaces after VLAN numbers on trunk ports,
  to allow long lists of VLANs on trunk ports to wrap when the
  window is resized.
Applied changes to improve support for HP switches.  Thanks to
  Nancy Greenblatt and Jeremy Lunsford.
Fixed UpdateOuiCodes.pl to handle the changed format of the IEEE
  oui.txt file, and to be more resilient to such changes in the
  future.
Updated the oui.txt file with the latest version from the IEEE, and
  ran UpdateOuiCodes to update the OuiCodes.txt file.


# Version 13.0, Feb 5, 2013
Major version bump because I added code to read MAC and VLAN data from
  Q-Bridge MIBs.  This enables SwitchMap to handle devices from
  vendors other than Cisco.  I've tested it on Juniper EX2200, EX8200,
  MX480 and QFX3500s.  Unfortunately, SwitchMap doesn't detect
  trunking correctly on some of these devices.  Baby steps :-)
Added GetMacsFromQBridgeMib.pm.
Refactored MAC-fetching code out of PopulatePorts.pm into
  GetMacsFromCiscoMibs.pm.  Still leaves over 1000 lines in
  PopulatePorts.pm.  Oh well.
Added WritePLookupDirectory.pm, which isn't useful outside my site.
Added GetVlansOnPorts.pm and some other code to show VLANs that are
  defined on trunk ports.  Very nice, thanks to Guus Houtzager.
Increased the buffer size on SNMP calls by adding
  "-maxmsgsize => 5000" to calls to Net::SNMP->session.  Thanks
  to spegghy69.
In SwitchUtils::GetWhatViaCdp, add 2 lines to delete imbedded
  newlines and carriage returns from the strings learned via CDP.
  Thanks to PCL.
Modified PopulatePorts.pm and a few others to handle switches that
  support the Cisco STACK MIB, but only partially.  This means the
  code now handles Cisco ME3600s.  Thanks to Imran Malik.

# Version 12.5, Jun 26, 2012
Changed Stats.pm so that it doesn't count ancillary ports towards
  the "active ports" total.  This lowers the number of active ports
  reported on the Statistics page.
Changed Stats.pm to add a new table named "Switch Details".  It
  shows the active, inactive, disabled and total port counts for
  each individual switch.  Handy for explaining where the counts
  in the "Port by State" table came from.
Changed GetIsPortTrunking to fix trunk detection of port channels
  on 3560s, which can have inconsistent "admin" and "oper" tables.
Changed GetIsPortTrunking to fix bug about "admin" vs. "oper"
  tables.  Thanks to Wolfgang Steudel.
Changed FindOffice.pl to make it correctly recognize MAC addresses
  of the form nnnn.nnnn.nnnn
Fixed a syntax bug in WritePoePortsFile.pm, in
  BuildApprovedManufacturersTable.  Thanks to Dan Thorson.

# Version 12.4, February 22, 2012
Changed PopulatePorts.pm so that if a switch has CDP disabled, the
  code won't skip the switch entirely.  Thanks to Isendel Steel.

# Version 12.3, February 9, 2012
Lots of changes to make better use of CSS.  Changed the alignment of
  some columns from "center" to "left", show MAC addresses in a
  fixed-width font, fix some redundancies, other stuff.
Made SwitchMap filter carriage-returns and line-feeds out of
  port label fields.  Some switches have CRs in the label fields,
  which screws up FindOffice.pl.  Thanks to DidierJ.
Added $ShowCdpNames to ThisSite.pm.  Thanks to Dan Thorson.
Added some port types to the Virtual Ports list.
When a routed port is inactive or disabled, show the VLAN number
  instead of 'routed'.  Thanks to Petr Cech.
Fixed problem with some switches reporting of
  pethPsePortDetectionStatus.
Changed the way CSV files are written.  Now, CSV files get the
  information for all ports, including inactive and trunk ports.
  There is a new variable in ThisSite.pm named
  $ShowOnlyActiveNonTrunkPortsInCsv that you can set to restore the
  old behavior.
Added a new option for specifying approved PoE devices.  Now you can
  use the %ManufacturersApprovedForPoe hash in ThisSite.pm to specify

ThisSite.pm
package ThisSite;

#
# This package defines site-specific constants used by the SwitchMap
# programs.
#

# Set $GetSwitchListFromHpOpenView to 1 (true) if you have HP OpenView
# Network Node Manager (NNM) running on a machine at your site, and
# you want SwitchMap to get the list of your switches from OpenView.
# This is a good thing to do if you can, because as you add/remove
# switches from your network, SwitchMap automatically adjusts.  If you
# leave $GetSwitchListFromHpOpenView set to 0 (false), SwitchMap will
# use the static @LocalSwitches array defined below.
$GetSwitchListFromHpOpenView = 0;

# Set $GetMacIpAddrFromHpOpenView to 1 (true) if you have HP OpenView
# Network Node Manager (NNM) running on a machine at your site, and
# you want SwitchMap to get the MAC-to-IP mapping tables from
# OpenView.  This is a good thing to do if you have OpenView and you
# have OpenView configured to manage all your hosts on all your
# networks.  Otherwise, leave $GetMacIpAddrFromHpOpenView set to 0
# (false) and SwitchMap will get the data from the MacList file (see
# the README file for an explanation of the MacList file)
$GetMacIpAddrFromHpOpenView = 0;

# Set $GetSnmpCommunitiesFromHpOpenView to 1 (true) if you have HP
# OpenView Network Node Manager (NNM) running on a machine at your
# site, and you want SwitchMap to get SNMP community strings from
# OpenView.  If you have OpenView and you have configured OpenView
# with SNMP community strings, setting this variable to true lets you
# avoid maintaining 2 lists of SNMP community strings.  Otherwise,
# leave $GetSnmpCommunitiesFromHpOpenView set to 0 (false) and see the
# comment that describes the $Community variable below.
$GetSnmpCommunitiesFromHpOpenView = 0;

# If you have HP OpenView Network Node Manager at this site (any of
# the previous 3 variables are set to true), then this program will
# get information from NNM.  To do so, this program needs to know the
# DNS name of the machine that is running HP OpenView NNM.  If the
# SwitchMap programs are running on the same machine as HP OpenView
# NNM, set $OpenViewHost to 'localhost'.  If HP Openview NNM is
# running on another machine, put the name here, and this program will
# ssh to the machine.  For this to work, you have to have ssh access
# to the machine defined by $OpenViewHost such that the user that runs
# the SwitchMap scripts is able to ssh to $OpenViewHost without
# supplying a password.  If the previous 3 variables are set to 0
# above, then it doesn't matter what value $OpenViewHost has.
$OpenViewHost = 'nnm.your.domain';

# If you have HP OpenView Network Node Manager at this site and NNM is
# running on another host, then this is the ssh key to use to get the
# switch lists and MAC data and the SNMP community strings file.
# SwitchMap will use this string as the argument of the "-i" option
# when SwitchMap does ssh or scp commands to get data from the remote
# host.
$SshKeyOption = '';

# If you do not have HP OpenView Network Node Manager at this site
# ($Has_HP_OpenView = 0 above), then you'll need to run GetArp.pl
# periodically to get IP and MAC data from your routers.  The @routers
# array lists the routers that the GetArp program will query.
@routers = ();
push @routers, 'router1';
push @routers, 'router2';

# @LocalSwitches provides a static list of switches that's used when
# you don't have HP OpenView ($GetSwitchListFromHpOpenView is set to
# 0).  If you have $GetSwitchListFromHpOpenView set to 1, SwitchMap
# ignores the contents of @LocalSwitches.  This list is used by
# ScanSwitch.pl and SwitchMap.pl.  When the programs run, switches
# will be accessed in the order that they appear in this list.  This
# is also the order that the switches will appear in the main
# "Switches" web page created by SwitchMap,pl.  The list can contain
# just switch names, but you can add special "title" cells.  Title
# cells are extra cells that dress up the switch list.  They are
# typically used to give a name (such as a building name) to a list of
# switch names.  A title cell starts with three dashes to
# differentiate it from normal switch names.  These "switch names" are
# ignored by most of the SwitchMap code, and only have meaning when
# the switches/index.html file is generated.  At that time, the cells
# are output as part of the list of switches, but with a different
# font.  A sample of how you might define them is shown here.  If you
# use title cells, you may be interested in the ReformatSwitchNames
# function described below.
@LocalSwitches = ();
push @LocalSwitches, '---Building1';
push @LocalSwitches, 'switch1-in-building1.abc.com';
push @LocalSwitches, 'switch2-in-building1.abc.com';
push @LocalSwitches, '---Building2';
push @LocalSwitches, 'switch1-in-building2.abc.com';
push @LocalSwitches, 'switch2-in-building2.abc.com';

# @LocalSwitchTrunkPorts provides a static list of trunk ports on
# switches.  It allows you to work around an unfortunate problem - on
# some switches, like Junipers and Cisco 3845s, SwitchMap can't tell
# the trunk ports apart from the non-trunk ports using SNMP.  For such
# switches, SwitchMap will creates web pages that show the trunk ports
# with all the MAC addresses that are reachable on the port - often
# hundreds of addresses.  This clutters the output pages and makes
# searching produce too many hits. I opened a case with Juniper
# support, but they were unable to provide a solution.  So I added the
# @LocalSwitchTrunkPorts array to SwitchMap version 13.1.  The array
# provides a way for you to explicitly name the trunk port(s) on your
# switches.  This is an inelegant, manual solution, and is vulnerable
# to getting out-of-date as you add/move/change your switches and
# trunk ports, but hey, it works.  SwitchMap checks the array when it
# assigns trunk status to ports.  If it finds a match, the port is
# marked as a trunk port.  If it doesn't find a match, SwitchMap tries
# it's usual algorithm for getting the trunk status.  You don't need
# to add all your trunk ports to the array, just add entries for your
# problem switches.
%LocalSwitchTrunkPorts = (
               'switch1'  => [ "ge-0/1/0", "ge-0/1/1" ],
               'switch2'  => [ "xe-0/0/1 ],
             );

# If you have HP OpenView and you've set $GetSwitchListFromHpOpenView
# to 1, it may match some switches that you don't want to appear in
# the port lists.  For example, at my site we have a network
# connection to a Catalyst that we monitor with HP OpenView even
# though we don't have administrative control of the switch.  So when
# SwitchMap finds all the Catalysts known to HP OpenView, it finds the
# switch along with all our other switches.  By putting that switch in
# the @SkipTheseSwitches list, we can make SwitchMap skip that switch.
@SkipTheseSwitches = ('ithaka-router');

# If you use the same community string for all your routers and
# switches, leave $CmstrFile as an empty string and set the $Community
# variable found below.  If you use different community strings in
# different devices, set $CmstrFile to the full pathname of a file
# that defines the strings.  The file should contain one community
# string per line, in double quotes.  Duplicates are ignored, as are
# lines that start with '#'.  On each line of the file, everything
# after the second double-quote character is ignored.  This format is
# identical to that of the netmon.cmstr file used by HP OpenView
# Network Node Manager, so that sites that have NNM can simply use the
# existing file.  If you have HP OpenView, set $CmstrFile to the full
# path name of the netmon.cmstr file, including the 'netmon.cmstr' at
# the end, and this program will try to open the file on the HP
# OpenView machine defined by $OpenViewHost above.  Specify the real
# full path name, without using any environment variables like
# OV_CONF.  If you don't have HP OpenView, set $CmstrFile to the full
# pathname of a file.  For each switch, the programs will try the
# community strings defined in the file, one after the other.  When a
# working community string is found, the programs use another file to
# save the string that worked, so that on subsequent runes they can
# avoid waiting while incorrect communities.
$CmstrFile = '';

# If you use the same SNMP community string in all your switches, set
# $Community to that value.  If your switches have different community
# strings, set the $CmstrFile variable above.  When that variable is
# set to a non-empty value, SwitchMap ignores the value of the
# $Community variable.  Note: the SwitchMap programs do only "get"
# SNMP requests - no SNMP "set" requests are done.
$Community = 'public';

# Your DNS domain.  A typical switch in our network is "abc.ucar.edu",
# so we set this to '.ucar.edu'.
$DnsDomain = '.your.domain';

# The $DestinationDirectory is the full path to the directory where
# the output files will be written.  This should be somewhere that
# your web server can access the files.  At my site, I set this to
# '/usr/web/nets/internal/portlists'.
$DestinationDirectory = '';

# The $DestinationDirectoryRoot is the path to the same directory as
# DestinationDirectory, but from the perspective of the web server.
# At my site, it's '/nets/internal/portlists'.
$DestinationDirectoryRoot = '';

# The $StateFileDirectory is the place that the programs will write
# and read the files that maintain state information from one run to
# the next.  These include the MacList file and the directories named
# 'idlesince' and 'community'.  This should be a different place than
# $DestinationDirectory, because the 'community' directory contains
# SNMP community strings, and we don't want web users to be able to
# browse to those files.  You can be lazy and make $StateFileDirectory
# the same as $DestinationDirectory.  This will work, but for security
# the programs will detect that you have and skip the use of the
# 'community' directory, and the programs won't remember working
# community strings, and may run slower.  On Unix systems, the
# standard place for state files is /var, so a good value is
# '/var/local/switchmap'.  You'll have to create the directory and set
# its ownership.  The directory must be writeable by the users that
# run the SwitchMap program.
$StateFileDirectory = '/var/local/switchmap';

# Set HasConfRooms to 1 if you have a static webpage named
# conference-rooms.html that describes your site's conference rooms.
# Setting it to 1 will cause SwitchMap to include a link to the page
# when it creates the portlists index page.  NCAR has such a page.
# You're not likely to have it, so you should probably leave
# $HasConfRooms set to 0.
$HasConfRooms = 0;

# The $WebPageTrailer variable contains a site-specific trailer that
# SwitchMap will put at the bottom of each web page.  What's here now
# will work, but to improve it I suggest you uncomment the next 5
# lines, put your email address in the obvious spot, and delete the
# 6th line.  my $YourEmailAddress = 'yourname@yourcampany.com';
# $WebPageTrailer = <<TRAILER; Address comments or questions about
# this web page to <a
# href="$YourEmailAddress">$YourEmailAddress</a>.<br> TRAILER
$WebPageTrailer = '';

# The $ExtraHelpText string contains extra site-specific text that is
# written to the help file.  The help file explains how the search
# function is used.  At my site, I initialize this variable with text
# that explains how we use the "name" fields in Cisco switches at our
# site.  You can safely leave this empty.
$ExtraHelpText = '';

# The number of days past which a port is considered "unused".
$UnusedAfter = 60;   # days

# The $UseSysNames variable is a boolean meant to help sites where DNS
# isn't available for switch names.  If your site has DNS names
# defined for each switch (most sites do), leave this variable set to
# 0.  If you don't have DNS names for your switches (you've had to use
# IP address in the LocalSwitches array above), set it to 1 and
# SwitchMap will use list your switches using the SNMP sysName value
# stored in the switches themselves.  Of course, this assumes you've
# set the sysName value in each of your switches using the "set
# hostname" or "set system name" command.
$UseSysNames = 0;

# The list of approved Power-over-Ethernet devices.  If you don't have
# POE devices, leave the @DevicesApprovedForPoe array empty.
#
# This is for sites like mine, that have POE devices like IP phones,
# and have chosen for convenience to configure all their switch ports
# to deliver POE.  We don't want our users to plug in POE devices
# other than phones, because we don't want our switches to have to
# support the power drain of clocks or other POE gadgets.  So we have
# a list of approved POE devices, and SwitchMap generates a web page
# showing all ports that are delivering POE to unapproved devices.  At
# my site, the list of approved devices is
#
# @DevicesApprovedForPoe = (
#                        'AIR-AP350',
#                        'AIR-BR350',
#                        'cisco AIR-AP1121G-A-K9',
#                        'Cisco IP Phone 7910',
#                        'Cisco IP Phone 7945',
#                        'Cisco IP Phone 7960',
#                        'Cisco IP Phone 7965',
#                        'Cisco IP Phone 7970',
#                       );
#
# If you leave the @DevicesApprovedForPoe empty, SwitchMap won't check
# for rogue POE devices.
#
# If you want to use this feature, the best way is to define at least
# one approved device, which can be bogus, in the
# @DevicesApprovedForPoe array.  Then run SwitchMap and see what you
# get in the POE web page, which you'll find under the "Ports" web
# page.  Then decide which of the devices ought to be approved, and
# add the value from the "What (via CDP)" column to the
# @DevicesApprovedForPoe.  When you're done, the POE page should show
# all the devices except the ones that you've decided are "approved".
#
@DevicesApprovedForPoe = ();

# The %MacsApprovedForPoe hash works like @DevicesApprovedForPoe, but
# lets you specify a list of specific MAC addresses that are approved.
# We found that we needed this because when some POE devices (like
# security cameras made by Topview or ACTi) negotiate with the switch
# for POE, they don't give the switch a device name.  To list them as
# "approved", I added this hash.  At my site, I initialize it like so:
#
# %MacsApprovedForPoe = (
#                       '000b67003be0'  => 'IP security camera made by Topview',
#                       '000b67004596'  => 'IP security camera made by Topview',
#                       );
%MacsApprovedForPoe = ();

# The %ManufacturersApprovedForPoe hash works like MacsApprovedForPoe,
# but allows you to specify manufacturers that supply MACs that you
# approve of.  This is more convenient than using %MacsApprovedForPoe,
# but may accidentally allow more devices than you intend.  It might
# be initialized like so:
#
# %ManufacturersApprovedForPoe = (
#                                '000b67' => 'IP security camera made by Topview',
#                                );
%ManufacturersApprovedForPoe = ();

# Some sites may want to improve the looks of the main switch list
# found in the index.html in the "switches" directory.  The following
# items ($UseGroups, WhichGroup and ReformatSwitches) provide ways to
# do that.  If you leave them at their default values, you'll get a
# simple alphabetized list of switch names.
#
# There are several styles you can use in your switch list:
#
# 1. The default, and simplest - just put switch names into the
#    @LocalSwitches list.  SwitchMap will write them to the web page
#    without changing the order, and split into the number of columns
#    defined by SwitchIndexColumns (defined below).
# 2. You can add "title cells" to the @LocalSwitches list.  These
#    strings start with three dashes.  SwitchMap ignores them when it's
#    processing switches, but adds them to the switch list web page, and
#    displays them in a bold font.  You might use these to put building
#    names into the switch list.  As with #1, you still maintain the
#    @LocalSwitches list by hand.
# 3. You can set the $UseGroups variable to 1 to "turn on" automatic
#    insertion of title cells.  This is useful if you can derive the
#    group a switch belongs to by the switch's name.  SwitchMap will
#    call the WhichGroup function to determine the group that each switch
#    belongs to, and add title cells for each group.  It's up to you to
#    define the WhichGroup function (see examples below).
# 4. You can define your own code in the ReformatSwitchNames function.
#    SwitchMap will call the function once, and pass in the
#    @LocalSwitches list by reference, so that you can modify it as
#    you like.  As delivered, ReformatSwitchNames does nothing to the
#    list.  You might use this to simply sort the @LocalSwitches list,
#    so you don't have to bother keeping the list in sorted order
#    yourself.  Or you might leave @LocalSwitches empty, and in
#    ReformatSwitchNames, fill the list with the contents of an
#    external file.

# If you want to split the list of switches in the "switches" index
# file into groups based on the switch names, set $UseRooms to "1",
# and change the code in the WhichGroup function below.
$UseGroups=0;

# If you have set UseGroups to "1" then you must tell Switchmap how to
# derive a group name from a switch name.  This routine takes the
# switch name as it's argument, and returns the group name.  This
# routine can be as simple or complex as needed, since we only call
# this routine once per switch.
# In the example code below, it is assumed that a typical switch name is
# of the form
#     building-room-switchtype-switchID.site.company.com
# and the example code returns
#     building-room
# as the group name.
sub WhichGroup($) {
  $_ = shift;
  # This example assumes names are XX-YY-anythingElse and so returns XX-YY
  # s/([0-9a-z]*)-([0-9a-z]*)-.*/$1-$2/;
  # return $_;

  # This example gets the portion of the domain name after the 1st dot
  # my @fqdn = split /\./;
  # return $fqdn[1];

  # This example returns the first four characters of the hostname
  # return substr $_, 0, 4;
}

#
# The ReformatSwitchNames subroutine is for SwitchMap users who want
# to do some site-specific modification of the @LocalSwitches list
# before SwitchMap uses it.  See the description above.
#
sub ReformatSwitchNames(@) {
  my $SwitchNamesRef = shift;
  # your code goes here to modify the switch name list
}

# The SwitchIndexColumns constant controls the number of columns that
# SwitchMap will use in the table that is created in the index.html
# file in the "switches" directory.
$SwitchIndexColumns = 5;

# When SwitchMap reads ARP caches to get MAC information, some
# interfaces may have too many MACs to display reasonably (like,
# hundreds).  We want to see the MACs on point-to-point interfaces,
# which often have only one MAC, but we don't want to see every MAC on
# crowded VLANs.  $ArpMacLimit is the maximum number of MACs to
# display from the ARP cache.  When there are more than this number of
# MACs in the ARP cache for a given interface, SwitchMap will display
# "<x> MACs in ARP cache, display limit is <n>".
$ArpMacLimit = 10;

# Ancillary ports are ports like "Nu0" (the null port) and "VLAN-nnn",
# which appear in the SNMP data structures in switches, but carry no
# information worth displaying.  To reduce clutter, SwitchMap gathers
# these ports into their own table.  The table is displayed only if
# the following boolean variable is set to true.  It's false by
# default.
$ShowAncillaryPorts = 0;

# If you set $ShowCdpNames to 1 (true), SwitchMap will put the CDP
# name into the "What (via CDP)" column.  This can be useful, but it
# can also make the column rather wide.
$ShowCdpName = 0;

# Historically, Switchmap wrote CSV files that contained only the
# information for active, non-trunk ports.  In 2011, some users
# requested that the CSV files should contain information for all
# ports, including inactive and trunk ports.  It is now the default
# behavior to generate CSV files containing information for all ports.
# If you prefer the old behavior, set
# $ShowOnlyActiveNonTrunkPortsInCsv to 1.
$ShowOnlyActiveNonTrunkPortsInCsv = 0;

# If you're on a Linux machine, you may have the arpwatch program
# installed.  If so, you can make the MacList file a bit more complete
# by feeding arpwatch information to GetApl.pl, which will then
# incorporate it into the MacList file.  See ReadArpWatchFile function
# in GetArp.pl.  Turn it on by setting $AprWatchFile to 1.  Thanks to
# pklausner.
$ArpWatchFile = 0;

# Sometimes it is useful to flag non-trunk ports that have too many
# MAC addresses.  Perhaps you need to hunt for user-installed hubs or
# switches at a site that has a policy against such things.  To make
# SwitchMap highlight such ports, set $PortMacLimit to a low number.
# For example, at my site, non-trunk switch ports are supposed to be
# connected a host, an IP phone, or a IP phone and host, so I'd set
# $PortMacLimit to 2.  The default value is "high", effectively
# turning off this feature of SwitchMap.
$PortMacLimit = 1000;

# The plookup (aka PetesLookup) program is a fast way to look up terms
# that are useful to network engineers (IP addresses, MAC addresses,
# etc.).  Set GeneratePLookupFiles to 1 to make SwitchMap generate
# files that can then be copied over to a PetesLookup repository for
# use by PetesLookup.
$GeneratePLookupFiles = 0;

1;
